internal/cmd/commands/dev/dev.go

```diff
@@ -824,20 +822,27 @@ func (c *Command) Run(args []string) int {
 	c.opsServer = opsServer
 	c.opsServer.Start()
 
-	// Wait for shutdown
-	shutdownTriggered := false
+	// On first interrupt, enter graceful shutdown for controllers and workers (workers wait for connections to drain)
+	// On second interrupt, force controller shutdown and send workers into shutdown (stop all connections)
+	// On third interrupt, force worker shutdown
+	gracefulShutdownTriggered := false
+	workerShutdownTriggered := false
 
 	// Add a force-shutdown goroutine to consume more interrupts
 	abortForceShutdownCh := make(chan struct{})
 	defer close(abortForceShutdownCh)
 
 	runShutdownLogic := func() {
 		go func() {
-			shutdownCh := make(chan os.Signal, 4)
-			signal.Notify(shutdownCh, os.Interrupt, syscall.SIGTERM)
 			for {
 				select {
-				case <-shutdownCh:
+				case <-c.ShutdownCh:
+					if !workerShutdownTriggered && !c.flagControllerOnly {
+						workerShutdownTriggered = true
+						if err := c.worker.Shutdown(); err != nil {
+							c.UI.Error(fmt.Errorf("Error shutting down worker: %w", err).Error())
+						}
```

Comment:
Keeping in mind that earlier the shutdown channel would be closed -- which is why it created a new channel here to listen for a second interrupt -- there is now a race condition between readers of `c.ShutdownCh`...depending on which one gets it first you may hit this case, or you may hit the case below and create a new goroutine running this code. Additionally when this is first run by the first struct down `c.ShutdownCh`, this case won't execute until a second interrupt comes in -- not sure if that's what you wanted or not.

There is also a data race on `workerShutdownTriggered` since more than one goroutine here can be launched concurently.

Additionally once `gracefulShutdownTriggered` in the case below the `for` loop will exit so you'll immediately return from this function.

Given the changes to `c.ShutdownCh`, I think what you may want to do is move the `gracefulShutdownTriggered` logic into `runShutdownLogic` so the for loop below is just a signal handler. You could instantiate an atomic uint32 that keeps track of how many interrupts have been received and is atomically incremented before `runShutdownLogic` is called; the function then switches logic based on that value. The first two interrupts trigger goroutines to either do a graceful shutdown or a non-graceful shutdown; the third runs `os.Exit`. Spitballing a bit there but I think something like that may be easier to reason about and avoid some of these signal handler issues.